llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -1845,7 +1845,39 @@ void DAGCombiner::Run(CombineLevel AtLevel) {
   DAG.RemoveDeadNodes();
 }
 
+static inline bool hasNoMergeProtection(const SDValue &V) {
+  if (SDNode *N = V.getNode()) {
+    if (N->getFlags().hasNoMerge())
+      return true;
+
+    // if V is (~X) expressed as (xor X, -1) also check X
+    if (V.getOpcode() == ISD::XOR) {
```

Comment:
This could block merging for all DAG paths. Could this have side-effect if someone create `XOR` node with operands all one const and expecting simplification of node?